table/block_based/filter_policy.cc

```diff
@@ -1107,14 +1242,54 @@ class AlwaysTrueFilter : public FilterBitsReader {
  public:
   bool MayMatch(const Slice&) override { return true; }
   using FilterBitsReader::MayMatch;  // inherit overload
+  bool HashMayMatch(const uint64_t) override { return true; }
+  using FilterBitsReader::HashMayMatch;  // inherit overload
+  bool IsAlwaysTrueFilter() override { return true; }
 };
 
 class AlwaysFalseFilter : public FilterBitsReader {
  public:
   bool MayMatch(const Slice&) override { return false; }
   using FilterBitsReader::MayMatch;  // inherit overload
+  bool HashMayMatch(const uint64_t) override { return false; }
+  using FilterBitsReader::HashMayMatch;  // inherit overload
 };
 
+Status XXPH3FilterBitsBuilder::PostVerify(const Slice& filter_content) {
+  Status s = Status::OK();
+
+  if (!detect_filter_construct_corruption_) {
+    return s;
+  }
+
+  std::unique_ptr<FilterBitsReader> bits_reader(
+      filter_policy_->GetFilterBitsReader(filter_content));
+  auto sync_point_callback_arg =
+      std::make_pair(&bits_reader, filter_content.ToString());
+  TEST_SYNC_POINT_CALLBACK(
+      "XXPH3FilterBitsBuilder::PostVerify::PreVerification",
+      &sync_point_callback_arg);
+
+  if (bits_reader->IsAlwaysTrueFilter()) {
+    // BuiltinFilterPolicy::GetFilterBitsReader() returns AlwaysTrueFilter for
+    // compability of filter content not supported by FastLocalBloomBitsReader
+    // nor Standard128RibbonBitsReader. However, such filter content should not
+    // be passed into XXPH3FilterBitsBuilder::PostVerify unless it experienced
+    // an corruption
+    s = Status::Corruption("Corrupted filter content");
```

Comment:
Note: I found it a lot cleaner to rely on `filter_policy_->GetFilterBitsReader(filter_content))` instead of manually creating FilterBitsReader according to the  filter_content, which duplicated some internal logic of `GetFilterBitsReader`. 

But there is one tricky point as described in comment in line 1274-1278 -`GetFilterBitsReader()` can `AlwaysTrueFilter()` for compatibility which defeats the purpose of detecting corruption in here. Therefore it asks for a way to differentiate `AlwaysTrueFilter` from other `FilterBitsReader` during run-time. 

To dd so and considering `RTTI` support can be unavailable according to `[..Replace dynamic_cast<> (except unit test) so people can choose to build with RTTI off. With make, release mode is by default built with -fno-rtti and debug mode is built without it. Users can override it by setting USE_RTTI=0 or 1.](https://github.com/facebook/rocksdb/blob/main/HISTORY.md#new-features-40)`, it seems that the only way left is to add some class info exposure method like `IsAlwaysTrueFilter()`. @mdcallag 's work in Customization in `FilterBitsBuilder` (if there is any) may also help in here.

Reply:
I haven't read all the details, but (in relation to #9412) it should be *safe* (though inefficient) to PostVerify an AlwaysTrueFilter, because all keys return `true` for MayMatch. I'm not sure it's worth carving out an exception for this obscure case.

Reply:
(no need to reply)Update: discussed offline to clarify the potential cost of ignoring "corruption to AlwaysTrueFilter/LegacyBloomBitsReader" case and we agreed that the cost is acceptable with the benefit of simplicity w/o special case handling and customization #9412